### DISCUSSION TITLE
Welcome to lippert-canbus Discussions!

### DISCUSSION THREAD

**Discussion by D-Jeffrey** [D-Jeffrey/UnityX-canbus#1] [Announcements]
## How does this work, let's figure it out together
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Share what you have learned by studying your Unity X CANBUS
  * Share what you have decoded and ways that we could reverse this bus
  * It is a Puzzle we will solve together

Let the old school (hacking)[https://hackaday.com/2013/10/21/can-hacking-introductions/] begin.

**Comment by ytechie** [comment]
Just curious if you ever made more progress. I have a similar controller and would love to interface with it.

**Reply by D-Jeffrey** [reply]
No I gave up on the approach of trying to decode Lipperts canbus.  My new approach was to buy a **Lippert Gateway** and use to create connect to the CANbus and make a OpenHAB instance which I can connect to with Home Assistant.  Overtime, I will revisit sniffing the command that OpenHAB generates (but since that thinking I have learned how complex the code is).  

Some of what I could not figure out turned out to be CRC related and the bus is using variable length addresses and messages ...

```
 public CAN_ID(IDS.Core.CAN.CAN_ID id)
    {
      if (id.IsExtended)
      {
        this.SourceAddress = (ADDRESS) (byte) ((uint) id >> 18);
        this.TargetAddress = (ADDRESS) (byte) ((uint) id >> 8);
        this.MessageData = (byte) (uint) id;
        this.MessageType = (MESSAGE_TYPE) (byte) (128 | (int) ((uint) id >> 24) & 28 | (int) ((uint) id >> 16) & 3);
      }
      else
      {
        this.SourceAddress = (ADDRESS) (byte) (uint) id;
        this.MessageType = (MESSAGE_TYPE) (byte) ((uint) id >> 8 & 7U);
        this.TargetAddress = ADDRESS.BROADCAST;
        this.MessageData = (byte) 0;
      }
    }
... and ... 
return this.Manager.Adapter.LocalHost.TransmitRequest((REQUEST) (byte) 38, (IBusEndpoint) this.Manager.Device, (byte) ((uint) (ushort) this.Manager.CurrentBlock >> 8), (byte) (ushort) this.Manager.CurrentBlock, (byte) ((uint) this.Manager.BlockDownloadFullCRC >> 24), (byte) ((uint) this.Manager.BlockDownloadFullCRC >> 16), (byte) ((uint) this.Manager.BlockDownloadFullCRC >> 8), (byte) (uint) this.Manager.BlockDownloadFullCRC);
```

I have had mixed sucess trying to decode the CANbus but not enough to come close to what I needed.  I would rather have a CANbus chip on a Pi, running Home Assistant, less parts but it appears that Lippert like other companies out there like Chamberlain want you to buy their custom software and hardware at a marked up retro code (while connecting it to their cloud at a extra cost).  I understand they have to pay their developers, but I wish they would support a open connection service.  For now I will have to lean on OpenHAB as a bridge.  There is a lot of code (most of what I don't need for my trailer).  

So unless someone else has another way forward, I will be stuck this method.

X180T <-> CanBus <-> Open Hab <-> Wired network <-> Custom OpenHab Intergration <-> Home Assistant

That might not work so Plan C is:
X180T <-> CanBus <-> Open Hab <-> MQTT bind <-> Wired network <-> MQTT <-> Home Assistant
- which is not great for me since I'm not a MQTT guy at all

Either way, I might have to build a lot of sensors for HA to pull the OpenHAB in.

So Plan A would still be:
X180T <-> CanBus <-> Custom: Lippert-CanBus <-> Home Assistant

Plan D is just live with OpenHAB    `:-(`

**Reply by D-Jeffrey** [reply]
The X180T <-> MQTT to HA would leverage what Bob has done which is via a Gateway Modem and OpenHab. 

https://github.com/rperciaccante/HomeAssistant_OneControl_Discovery

Related information to this would be : https://github.com/rperciaccante/onecontrol2mqtt

**Reply by ytechie** [reply]
thanks - I seem to have a new/unusual controller (x270), and I have no idea if Grand Design customized it at all when they private labeled it as "Compass Connect". This is all probably more work to integrate than it's worth. That's really disappointing.

**Comment by D-Jeffrey** [comment]
It is very complicated, I do it as a puzzle to solve rather tha a few clicks to turn it on.

A lot of the equipment is made by Lippert.  https://support.lci1.com/electronics/ 

If you want to go down the rabbit hole then you can access their technion training at: https://lippert.learnupon.com/ 

They do have some course material which cover the X260 which I assume the X270 is the next model. has one high amperage relay than than the X270.  (You have more liquid sensors, a generator connection)  The core of the board is the same between the X180, X260 and X270.  
The [Lippert OneControl X Series Monitor Panels](https://lippert.learnupon.com/enrollments/176843708/page/686101753#) cover your model.  [reference for X-Series](https://lci-support-doc.s3.amazonaws.com/Training%20Hand-outs/onecontrol-x-series-master-training-powerpoint-hand-out.pdf)

Compass Connect goes by several names: 

- MyRV
- LCI OneControl
- Grand Design Compass Connect

(Off on a rabbit trail...)  Depend on your trailer manufactor most will use he OneControl approach.  Sounds like you have a Grand Design.  Other Manufactures are:
- Forest River (including Rockwood and Flagstaff)
- Thor (Airstream, Dutchmen, Heartland, Highland Ridge, Jayco, Starcraft, Keystone, KZ/Sportsman)
- Elkhart
- Winnebago (which is also Grand Design RV and Newmar)

Lippert runs with the brands of:

- Furrion
- Aries
- Curt
- Greystone
- HappiJac
- OneControl
- Solera
- SureShade
- Somnum
- Thomas Payne
- and many more

Over on https://gdrvowners.com/forum/operation/appliances you will find some limited tinkering.  
There are a few facebook groups who are hacking on their stuff, specifically **[DIY Smart RV](https://www.facebook.com/groups/diysmartrv/)** (Bob runs a good group of very leading edge), and many 'Grand Design' groups.

That maybe the place to get the answers you are looking for.

**Comment by ytechie** [comment]
Oh wow, that's way more information that I was able to dig up in my searches. Thank you - super cool. I also requested to join that Facebook group.

**Comment by pokemane** [comment]
I just stumbled on your project here after curiosity got the better of me and I cracked open the cabinet behind the panel. I have a 22BHE floorplan, 1 awning 2 slides, interior and exterior lights, 3 tanks (fresh/gray/black), 1 pump.

## Things I've learned so far

https://gist.github.com/pokemane/2b762b213cc9e87e1285678fb2fd77cf

### Read capabilities

I can read battery voltage (determined from watching sag during slide/pump/light operation), on/off/state of buttons and functions, light brightness, high-power channel current (the slide and awning and water pump current, figured this out because you can actually see the inrush for the pump and slides), and several candidates for tank fullness (haven't gotten to test these yet because I haven't filled them to check). 

Nothing is obfuscated. The hardest or most complicated part is the dynamic address negotiation on startup, which means that you can't hardcode the full message list like it's a DBC. You need to "learn" it based on an initial session of pushing all the buttons -> restarting (battery dc) -> then pushing all the buttons again to confirm address family codes. Details below. 

### Write capabilities [**safety notice**]

Write-path exists but is gated behind a multi-frame handshake that looks like a **safety interlock**. I am treating this as out-of-scope and readonly for now. 

**Something like this is there for a _very_ good safety reason..** I'm not touching it in this exercise. 

### Message ID layout
```
29-bit:
  bits  0..7   message data
  bits  8..15  TARGET address
  bits 16..17  message type, low 2 bits
  bits 18..25  SOURCE address
  bits 26..28  message type, high bits
  MessageType = 128 | ((id >> 24) & 28) | ((id >> 16) & 3)

11-bit:
  bits 0..7    SOURCE address (target implicitly broadcast)
  bits 8..10   message type
```

### 11-bit function codes
 
| fn | meaning |
|----|---------|
| 0 | ident — type + serial; bit2 of byte0 pulses 1 s after state change |
| 1 | always `00 00 00 00` so far — **still unexplained** |
| 2 | aux — stable channel fingerprint (see below) |
| 3 | status — state flags + 16-bit load current |
| 6 | single state byte |
| 7 | heartbeat, 1 Hz counter in byte 3 |


### Dynamic addressing
 
Node addresses are **reassigned every controller boot** via address-claim
frames on ID `0x000`: `<claimed_addr> <type> 00 00 00 <serial:3>`.
Nodes `0x01` and `0x04` (serials `2983A7`, `2983A8`, not the X180D) keep
static addresses.
 
Stable identity = **aux frame bytes [3:8]**. Byte 2 is the session master
address and changes with the epoch. Byte 3 is a device class:
`0x14` dimmer, `0x1E` switch, `0x21` motor, `0x27` analog, `0x0A` tank,
`0x2B` panel.
 
Claimed addresses observed across two boot epochs span nearly the full byte range (`0x0A`–`0xF0`, then `0x0D`–`0xF8`) — this does **not** follow the conventional J1939 "dynamic range" convention (typically `0x80`–`0xF0`). Looks more like a hash of device serial than a reserved-range allocation. Consequence: don't hardcode a source address for a custom node on the assumption it's "safe." Listen through the full boot address-claim sequence (ID `0x000`) first, collect what's actually claimed that epoch, and only then transmit a claim for something unclaimed.

### Status payloads (fn3)
 
- motor: byte0 bits0-1 direction (`C0` idle / `C3` active), bytes 2:3 current u16 BE
- switch: byte0 bit0 on/off (`80`/`81`), bytes 2:3 current u16 BE
- dimmer: byte0 bit0 on/off, bytes 1 and 3 level (duplicated), 0..255
- analog (battery): byte 2, u8
- tank: [still unconfirmed, but there are several tank-shaped message families that I need to follow up on]

Load current decodes as u16 BE (fn3 bytes 2:3), but the count-to-amp scale is not global... it differs by channel, almost certainly because the current-sense shunt is sized per channel rating. A single gain fit to the pump reads nonsensically on the slides. Each channel (or channel class) needs its own clamp-meter fit; the values in the code in the gist are unverified placeholders. I haven't taken a current clamp to any of the actual feed wires to confirm anything. 

The battery voltage is a little different, because I could watch it sag during some of my captures. I'm more confident in that, but not 100%.

**Reply by D-Jeffrey** [reply]
It seems you are working at the hardware level, which is beyond my skills.  I would direct you to look at the work of [@phurth](https://github.com/phurth/) and specifically his [list of constants](https://github.com/phurth/ha-onecontrol/blob/main/custom_components/ha_onecontrol/const.py) and others chunks of code such as [commands](https://github.com/phurth/ha-onecontrol/blob/main/custom_components/ha_onecontrol/protocol/commands.py).  

That is an interesting chunk of code.

**Reply by pokemane** [reply]
yeah i have outside experience with CAN so that part wasn't as scary to me as looking at bluetooth.
